File: git-ff.rb
#!/usr/bin/ruby

#   Simple fast-forward without working tree
#   ruby version

#   Advances the branch ref if it is and ancestror of target.
#   No need of a working tree as it does not try to merge but uses git
#   merge-base.
#
#   use an alias in .gitconfig:
#     ff   = !git-ff "$@"
#
#   author: piotr.lesnicki@ext.bull.net



def ancestor?(a, b)
  common = `git merge-base #{a} #{b}`
  return common == `git rev-parse #{a}`
end

def merge_status(ref, up)
  if `git rev-parse #{ref}` == `git rev-parse #{up}`
    return "up-to-date"
  elsif ancestor? ref, up
    n = `git rev-list #{ref}..#{up}`.lines.count
    return "behind #{n}"
  elsif ancestor? up, ref
    n = `git rev-list #{up}..#{ref}`.lines.count
    return "ahead #{n}"
  else
    return "merge/rebase required"
  end
end

def all_status()
    r = `git for-each-ref --shell --format='ref=%(refname); up=%(upstream);' refs/heads/`
    fail if $?.to_i != 0
    r.lines do |l|
      ref=nil; up=nil
      eval l
      if up != ""
        status = merge_status ref, up
        puts "#{'%-20s'%ref}: [#{up}: #{status}]"
      else
        puts "#{'%-20s'%ref}: not tracking"
      end
    end
end

def ff(branch)
  if branch == ""
    branch = `git rev-parse --abbrev-ref HEAD`
  end

  `git show-ref --heads -q #{branch}`
  if $?.to_i != 0
    raise "branch #{branch} is not a local head reference"
  end
  branch = `git show-ref --heads #{branch}`.lines.collect{ |l| l.split[1]}[0]

  branch_sha = `git rev-parse --verify -q #{branch}`
  if $?.to_i != 0; raise "error: branch #{branch} invalid" end

  branchname = branch.sub("refs/heads/","")
  upstream = `git rev-parse --symbolic-full-name --verify -q \
                #{branchname}@{upstream} 2>/dev/null`.strip
  if $?.to_i !=0; raise "error: not tracking anything" end

  status = merge_status branch, upstream
  puts "#{branch.sub("refs/","")} [#{upstream.sub("refs/","")}: #{status}]"
  if status.include? "behind"
    puts "  fast-forwarding %s..%s" \
    % [`git rev-parse --short #{branch}`.strip, `git rev-parse --short #{upstream}`.strip]
    `git update-ref #{branch} #{upstream}`
    if $?.to_i !=0; raise "error: update ref #{branch} to #{upstream} failed" end
  elsif status.include? "ahead"
    remote = upstream.sub("refs/remotes/","").split("/")[0]
    upref = upstream.sub("refs/remotes/","").split("/")[1..-1]
    puts "  you can push: git push #{remote} #{branch.sub("refs/","")}:#{upref}"
  elsif status.include? "up-to-date"
  elsif status.include "merge"
    puts "  merge/rebase required"
  else
    raise "  bad status"
  end
    
end

def ff_all()
  r=`git for-each-ref --shell --format='ref=%(refname); up=%(upstream);' refs/heads/`
  if $?.to_i !=0; raise "error: for-each-ref failed" end
  r.lines do |l|
    ref=nil; up=nil;
    eval l
    if up != ""
      ff ref
    end
  end
end


def main()
    if (ARGV.include? '-h' or ARGV.include? '--help')
      $stderr.puts "usage: <branch> || --all || --status"
      $stderr.puts "  fast-forward branch <branch>"
      exit 1
    end
    arg = ARGV[0]
    if arg == '--status'
      all_status
    elsif arg == '--all'
      ff_all
    elsif arg and arg != ''
      ff arg
    else
      puts "nothing"
    end
end


if __FILE__ == $0
  main
end

File: git-ff.py
#!/usr/bin/python

"""
git-ff [options] [<branch>]

 fast-forward <branch> (or by default current branch)

 fast-forward without trying to merge : just update the
 reference. Usefull especially without working directory.
 
 usage:
  -h | --help
  -s | --status  : show merge state of all branches
  -a | --all     : fast forward all branches
"""


from subprocess import Popen, PIPE, CalledProcessError
import sys
import getopt

## not finished

def sh(cmd):
    "call shell command, in python 2.6+. strip eol."
    p = Popen(cmd, shell=True, stdout=PIPE)
    if p.wait() != 0: raise CalledProcessError(p.returncode, cmd)
    return p.stdout.read().strip('\n')


def is_ancestor(a, b):
    "if commit a is ancestor of commit b"
    common = sh("git merge-base %s %s " % (a, b))
    return common == sh("git rev-parse %s" % a)

def merge_status(ref, up):
    "merge status with upstream : up-to-date | behind n | ahead n | merge required"
    if sh("git rev-parse %s"%ref) == sh("git rev-parse %s"%up):
        return "up-to-date"
    elif is_ancestor(ref, up):
        n = len(sh("git rev-list %s..%s"%(ref,up)).splitlines())
        return "behind %d" % n
    elif is_ancestor(up, ref):
        n = len(sh("git rev-list %s..%s"%(up,ref)).splitlines())
        return "ahead %d" % n
    else:
        return "merge/rebase required"

def all_status():
    "print merge status of all branches"
    refs = sh("git for-each-ref --shell --format='(%(refname),%(upstream))' refs/heads/")
    for l in refs.splitlines():
        ref, up = eval(l)
        if up != "":
            status = merge_status(ref, up)
            print "%-20s: [%s: %s]" % (ref, up, status)
        else:
            print "%-20s: not tracking" % ref


def git_ff(branch):
    "fast forward branch if needed (update the ref)"
    if branch is None:
        # take current branch
        branch = sh("git rev-parse --abbrev-ref HEAD")
    try: branch = sh("git show-ref --heads %s" % branch).split()[1]
    except: sys.exit("error: branch %s does not exist" % branch)
    branch_sha = sh("git rev-parse --verify -q %s"%branch)

    try: upstream = sh("git rev-parse --symbolic-full-name"
                  " --verify -q %s@{upstream} 2>/dev/null"
                  % branch[len("refs/heads/"):])
    except: sys.exit("error: no upstream for branch %s" % branch)

    status = merge_status(branch, upstream)
    print "%-20s %-25s : %-s" % (branch[len("refs/"):],
                                 upstream[len("refs/"):], status)
    if status.startswith("behind"):
        print "  fast-forwarding %s..%s" % (
            sh("git rev-parse --short %s"%branch),
            sh("git rev-parse --short %s"%upstream))
        sh("git update-ref %s %s" % (branch, upstream))
        return 0
    elif status.startswith("ahead"):
        remote, upref = upstream[len("refs/remotes/"):].split("/", 1)
        print "  you can push: git push %s %s:%s" %(remote, branch[len("refs/"):], upref)
        return 0
    elif status == "up-to-date":
        return 0
    else:
        return 1
    

def git_ff_all():
    "fast forward all tracking branches"
    refs = sh("git for-each-ref --shell --format='(%(refname),%(upstream))' refs/heads/")
    for l in refs.splitlines():
        ref, up = eval(l)
        if up != "":
            git_ff(ref)

def _test():
    print "faulty rev-parse:"
    try:
        print sh("git rev-parse --verify feb0f"),
    except CalledProcessError, e:
        print "rev-parse failed:", e

    print "all status:"
    all_status()
    print
    branch = "ruby"
    print "gitff %s" % branch
    git_ff(branch)

def main(argv):
    opts, args = getopt.getopt(argv, "hsa", ["help", "status", "all"])
    opts = dict(opts)
    if '-h' in opts or '--help' in opts:
        sys.exit(__doc__.strip())
    elif '-s' in opts or '--status' in opts:
        all_status()
    elif '-a' in opts or '--all' in opts:
        git_ff_all()
    else:
        branch = args[0] if len(args)>0 else None
        #branch = None
        git_ff(branch)

if __name__ == '__main__':
    #_test()
    main(sys.argv[1:])

    
